### PR TITLE
Bugfix  add delete to show Membership Application for admin 

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story: **Can't delete duplicated applications**
https://www.pivotaltracker.com/story/show/136390979

Changes proposed in this pull request:
1. Red Delete/Ta bort - link visible only to admin on "View membership application" page (used a link instead of a button just to keep things simple. (no underlying form needed for just the 1 button). Could be styled as a button, though.)
2. refactored the policies: moved `is_admin?` up to common parent (ApplicationPolicy)
3. added policy check for `MembershipApplicationController#destroy?`

Screenshot: 
skarmavbild 2016-12-20 kl 08 22 25

ready for review
@AgileVentures/shf-project-team

### CODE REVIEW

File: app/views/membership_applications/edit.html.haml [hunk 1]
```diff
@@ -13,3 +13,6 @@
           = msg
     = render partial: 'membership_application_fields', locals: {f: f}
     = render 'required_fields'
+
+    - if current_user.admin?
+      = link_to "#{t('.delete')}", @membership_application, method: :delete, data: { :confirm => "#{t('.confirm_are_you_sure')}" }
```
Comment (thesuss) [inline app/views/membership_applications/edit.html.haml]: was it { confirm: "#{t('.confirm_are_you_sure')}" that didnt work?

Comment (thesuss) [inline app/views/membership_applications/edit.html.haml]: data: { confirm: "#{t('.confirm_are_you_sure')}" } 
should work

File: features/step_definitions/membership_application_steps.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@
   visit edit_membership_application_path(membership_application)
 end
 
-Given(/^I am on "([^"]*)" application page$/) do |first_name|
+Given(/^I am on (the show )+"([^"]*)" application page$/) do |optional, first_name|
```
[the remark this answers is not in the document]
Comment (weedySeaDragon) [reply features/step_definitions/membership_application_steps.rb]: I know, houndbot.  But it's a regular expression and I don't know how to deal with it any better than that right now.

**Comment by thesuss** [comment]
I'll check this at breakfast break :)

File: app/controllers/membership_applications_controller.rb [hunk 3]
```diff
@@ -68,6 +68,11 @@ def information
 
   end
 
+  def destroy
+    @membership_application.destroy
+    redirect_to membership_applications_url, notice: t('membership_applications.application_deleted')
```
Comment (patmbolger) [inline app/controllers/membership_applications_controller.rb:73]: Do we need the full URL path or can we get away with a relative path (that is, use `membership_applications_path`)?

File: app/policies/application_policy.rb [hunk 4]
```diff
@@ -15,11 +15,23 @@ def index?
   end
 
   def update?
-    @user.admin? ||  @record.user == @user
+    is_admin? ||  @record.user == @user
```
Comment (patmbolger) [inline app/policies/application_policy.rb:18]: We could also just do `@user&.is_admin?` and dispense with the helper method.

(similarly for other policy files with the same helper method)

Comment (patmbolger) [reply app/policies/application_policy.rb:18]: Also, I notice that we have inconsistent instance var references in our policy files - that is, in some we refer to `@user` (like this one) and others we use the reader method `user` (like in `membership_application_policy.rb`).

We should be consistent in these references - I suggest the reader methods be used.

**Review by patmbolger — COMMENTED** [review]
Reviewed .... please see comments.

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:73]: You should use URLs with `redirect_to` : http://api.rubyonrails.org/classes/ActionController/Redirecting.html#method-i-redirect_to

The general guideline is:  `_url` in controllers (think: server side),` _path` in views (think: browser side)

File: app/policies/application_policy.rb (hunk 4 above)
Comment (weedySeaDragon) [reply app/policies/application_policy.rb:18]: totally agree. policies need cleaning up/refactoring.  we have a chore about refactoring MembershipApplicationPolicy, but we really do need to do all of them.  I'll make one.

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (patmbolger) [inline app/controllers/membership_applications_controller.rb:73]: I noticed that the english wording (in yml file) for this is: ```Application deleted.```, yet the Swedish is ```Ansökan raderad.``` - which is plural ("Applications deleted.") and should be singular.

May want to fix in this PR, as it affects new tests in this PR.

**Review by patmbolger — CHANGES_REQUESTED** [review]
Review - please see additional comment.

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:73]: (changed my mind again again.)

I'll fix it.

**Comment by weedySeaDragon** [comment]
suddenly, the build is failing here, but passing on my local system.  
So I'm reverting until I can get back to a passing build. (I suspect it's the automatic merge done by github)

**Comment by weedySeaDragon** [comment]
Not entirely sure what was going on with my local branch vs. GH.  But it's passing.

ready for another review @patmbolger

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: None of them are plural

**Comment by thesuss** [comment]
There is an intermittent error in the build. That's why I pushed for the refactoring as prioritized.

**Comment by weedySeaDragon** [comment]
Which branch did you first see the intermittent error?  
I wouldn't describe the build error that I got here as 'intermittent' -- it was definitely consistent.  So (as you can all see), I reverted my commits and re-did the changes.  I left off moving the info in the locale files.  (Plus I was tired and needed to take a break.)

Yes -- definitely time to refactor and clean up.  (Even if we didn't have intermittent errors, it's needed.)

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:73]: We can fix the locale files for that in a tiny PR after we get all of the sections completed with the PivT feature https://www.pivotaltracker.com/story/show/133316647

I'd really like to get the feature done -- touching all of the models, controllers, features, locale files.   :-)

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: I don't know why @patmbolger trusts Google translate over me?

**Comment by weedySeaDragon** [comment]
ready for review (finally) @patmbolger 
(figured out that the intermittent failing was due to the wrong db_cleaning strategy being used for cucumber tests.  I changed that: should have been _truncation_).

Susanna can make sure the text is right in the locale files in another PR if needed.  Now the infrastructure is there _and_ the features and controller are using the t() as appropriate.

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:73]: @thesuss - I *don't* trust google translate (GT) over you - it's just that it is the only way I can test English-to-Swedish translations.

Your judgement is the only one that counts :)

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: The swedish was there (hardcoded in the views/messages), its the english that is added

Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:73]: @thesuss  - yes, indeed .... and so I will focus on checking the swedish-to-english translations.

In this case, google translate is wrong.

However, it looks like rails i18n implementation is smarter than GT.  When I run this branch on my machine (with locale == english), the translated flash message is correct:

screen shot 2016-12-22 at 5 50 37 am

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: Yes, but it was the swedish one you said something about, is that still correct?
Answer is no:
`application_deleted: Ansökan raderat.`
It was correct, thus should be
`application_deleted: Ansökan raderad.`

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: But, that could be fixed separately, if someone *hint* puts it in piv.

**Review by patmbolger — APPROVED** [review]
PR looks good.

File: app/controllers/membership_applications_controller.rb (hunk 3 above)
Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:73]: Not sure I follow this comment ... do we need to change the translation?  But, you say "it was correct ..." so I'm not sure what is meant here.

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: it *WAS* correct, when you added this comment. Then Ashley changed it. Now it is not correct. =) 
it is (now)
`application_deleted: Ansökan raderat.`
It was, and should be
`application_deleted: Ansökan raderad.`

Comment (thesuss) [reply app/controllers/membership_applications_controller.rb:73]: I added it here: https://www.pivotaltracker.com/story/show/133316647

Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:73]: @weedySeaDragon - do you want to undo your change to fix this in this PR?

**Pull request merged**